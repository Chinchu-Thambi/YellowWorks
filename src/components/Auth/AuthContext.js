/* globals localStorage */

import React, {
  createContext, useReducer,
} from 'react';
import { navigate } from 'gatsby';
import { Hub } from '@aws-amplify/core';
import axios from 'axios';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import { NotificationManager } from 'react-notifications';

import useModule from '../../services/useModule';
import { storageAvailable, useLocalStorage } from '../../util';
import clearSubscriptionFormData from '../../util/clearSubscriptionFormData';
import { segmentIdentifyAndTrack } from '../../util/segment';

import getCustomerId from './services/getCustomerId';
import getStaffStatus from './services/getStaffStatus';

const initialState = {
  status: 'AUTHENTICATING',
  user: null,
  modal: {
    showAuthModal: false,
    // The target URL when this popup show up. Usually happens when accessing manage page directly
    params: null,
    target: null,
    dismissTarget: null,
  },
  customerId: null,
};

const reducer = (state, {
  type, user, error, modal,
  payload,
}) => {
  switch (type) {
    case 'AUTHENTICATED':
      return {
        ...state,
        status: 'AUTHENTICATED',
        user,
        modal: modal !== undefined ? modal : state.modal,
      };
    case 'UNAUTHENTICATED':
      return {
        ...state,
        status: 'UNAUTHENTICATED',
        error: error !== undefined ? error : state.error,
        modal: modal !== undefined ? modal : state.modal,
      };
    case 'AUTHENTICATING':
      return {
        ...state,
        status: 'AUTHENTICATING',
        modal: modal !== undefined ? modal : state.modal,
      };

    case 'LOGOUT':
      return {
        ...state,
        status: 'UNAUTHENTICATED',
        user: null,
      };

    case 'REGISTRATION_ERROR':
      return {
        ...state,
        error: error !== undefined ? error : state.error,
      };

    case 'SELECT_CUSTOMER':
      return {
        ...state,
        user: {
          ...state.user,
          selectedCustomer: payload.selectedCustomer,
        },
      };
    case 'TOGGLE_MODAL':
      return {
        ...state,
        modal: {
          ...state.modal,
          ...modal,
        },
      };
    default:
      return state;
  }
};

const AuthContext = createContext(null);

export default AuthContext;

let authState;

export const AuthProvider = ({ children }) => {
  const { Auth } = useModule('@aws-amplify/auth') || {};
  const [state, dispatch] = useReducer(reducer, initialState);
  const [selectedCustomer, setSelectedCustomer] = useLocalStorage('selectedCustomer', null);

  const jwtToken = R.path(['user', 'signInUserSession', 'idToken', 'jwtToken'])(state);
  const payload = R.path(['user', 'signInUserSession', 'idToken', 'payload'])(state);
  const expiryTime = R.path(['user', 'signInUserSession', 'idToken', 'payload', 'exp'])(state);
  const isStaff = getStaffStatus(state);
  const customerId = getCustomerId(state);

  // Gets current authentication state
  const getCurrentUser = async ({ bypassCache = false } = {}) => {
    try {
      dispatch({ type: 'AUTHENTICATING' });
      const currentUser = await Auth?.currentAuthenticatedUser({ bypassCache });

      dispatch({
        type: 'AUTHENTICATED', user: currentUser, modal: { showAuthModal: false },
      });
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(error);
      dispatch({ type: 'UNAUTHENTICATED', error });
    }
  };

  const refreshUser = React.useCallback(() => {
    if (!Auth) {
      return;
    }
    getCurrentUser({ bypassCache: true });
  }, [Auth]);

  const register = async (username, password, firstName, lastName) => {
    try {
      await Auth.signUp({
        username,
        password,
        attributes: {
          given_name: firstName,
          family_name: lastName,
        },
      });
      return true;
    } catch (error) {
      // TODO handle login error
      // eslint-disable-next-line no-console
      console.log(error);
      dispatch({ type: 'REGISTRATION_ERROR', error });
      return false;
    }
  };

  const confirmSignUp = async (username, code) => {
    try {
      const result = await Auth.confirmSignUp(username, code);
      return { success: true, result };
    } catch (error) {
      return { success: false, error };
    }
  };

  const resendCode = async (username) => {
    try {
      await Auth.resendSignUp(username);
      return { success: true };
    } catch (error) {
      // TODO handle login error
      // eslint-disable-next-line no-console
      console.log(error);
      return {
        success: false,
        error,
      };
    }
  };

  const login = async (username, password) => {
    try {
      const user = await Auth.signIn(username, password);
      dispatch({ type: 'AUTHENTICATED', user, modal: { showAuthModal: false } });
      const {
        attributes: {
          sub: id, email, given_name: givenName, family_name: familyName,
        },
      } = user;
      segmentIdentifyAndTrack({
        event: 'Login',
        id,
        payload: {
          email,
          firstmame: givenName,
          lastname: familyName,
        },
      });
      return { success: true, user };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = async () => {
    try {
      const { user } = state;
      if (user) {
        const {
          attributes: {
            sub: id, email, given_name: givenName, family_name: familyName,
          },
        } = user;
        segmentIdentifyAndTrack({
          event: 'Logout',
          id,
          payload: {
            email,
            firstname: givenName,
            lastname: familyName,
          },
        });
      }
      await Auth.signOut();

      clearSubscriptionFormData();

      navigate('/');
    } catch (error) {
      // TODO generic error handler
      // eslint-disable-next-line no-console
      console.log(error);
    }
    dispatch({ type: 'LOGOUT', modal: { showAuthModal: false } });
  };

  const forgotPassword = async (username) => {
    try {
      const result = await Auth.forgotPassword(username);
      return result;
    } catch (error) {
      // TODO generic error handler
      // eslint-disable-next-line no-console
      console.log(error);
      return { error };
    }
  };

  const forgotPasswordSubmit = async (username, code, newPassword) => {
    try {
      await Auth.forgotPasswordSubmit(username, code, newPassword);
      return true;
    } catch (error) {
      // TODO generic error handler
      // eslint-disable-next-line no-console
      console.log(error);
      return { error };
    }
  };

  const changePassword = async ({ previousPassword, proposedPassword }) => {
    const response = await Auth.changePassword(
      state.user,
      previousPassword,
      proposedPassword,
    );

    return response;
  };

  const loginGoogle = () => {
    if (storageAvailable()) {
      localStorage.setItem('yp_redirect', R.path(['modal', 'target'])(state));
      localStorage.setItem('yp_provider', 'Google');
      Auth.federatedSignIn({ provider: 'Google' });
    }
  };

  const loginFacebook = () => {
    if (storageAvailable()) {
      localStorage.setItem('yp_redirect', R.path(['modal', 'target'])(state));
      localStorage.setItem('yp_provider', 'Facebook');
      Auth.federatedSignIn({ provider: 'Facebook' });
    }
  };

  const loginYellow = () => {
    if (storageAvailable()) {
      localStorage.setItem('yp_redirect', R.path(['modal', 'target'])(state));
      localStorage.setItem('yp_provider', 'Yellow');
      Auth.federatedSignIn({ provider: 'Yellow' });
    }
  };

  const showHideModal = ({
    show, modalType, target, params, dismissTarget, dismissable,
  }) => {
    dispatch({
      type: 'TOGGLE_MODAL',
      modal: {
        modalType, showAuthModal: show, target, params, dismissTarget, dismissable,
      },
    });
  };

  const displayLoginModal = ({ target, dismissTarget, dismissable } = {}) => {
    showHideModal({
      show: true,
      modalType: 'login',
      target,
      dismissTarget,
      dismissable,
    });
  };

  const getCustomerListByEmail = async (email) => {
    const { data: { data, errors } } = await axios.post(
      process.env.GATSBY_APPSYNC_URL,
      {
        query: `query {
          getCustomerListByEmail(email: "${email}"){
            name
            companyId
          }
        }`,
      },
      {
        headers: {
          Authorization: jwtToken,
        },
      },
    );

    if (errors) {
      // eslint-disable-next-line no-console
      console.error('ERRORS', errors);
      throw errors;
    }

    return data;
  };

  const selectCustomer = React.useCallback(async (customerObject) => {
    if (!customerObject || !isStaff) return;

    const { contactId, customerId: selectedCustomerId, customerName } = customerObject;

    setSelectedCustomer({
      contactId,
      customerId: selectedCustomerId,
      customerName,
    });
  }, [isStaff, setSelectedCustomer]);

  // Configure Auth
  React.useEffect(() => {
    if (!Auth) {
      return;
    }

    Auth.configure({
      region: `${process.env.GATSBY_AWS_REGION}`,
      userPoolId: `${process.env.GATSBY_COGNITO_USER_POOL_ID}`,
      userPoolWebClientId: `${process.env.GATSBY_COGNITO_POOL_CLIENT_ID}`,
      oauth: {
        domain: `${process.env.GATSBY_OAUTH_DOMAIN}`,
        scope: ['email', 'profile', 'openid', 'aws.cognito.signin.user.admin'],
        redirectSignIn: `${process.env.GATSBY_OAUTH_REDIRECT_SIGN_IN}`,
        redirectSignOut: `${process.env.GATSBY_OAUTH_REDIRECT_SIGN_OUT}`,
        responseType: 'code',
      },
    });
  }, [Auth]);

  // Gets initial authentication state
  // Second argument of empty array means this effect will only run once
  React.useEffect(() => {
    if (!Auth) {
      return;
    }

    getCurrentUser();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [Auth]);

  // Redirects user post social login
  React.useEffect(() => {
    if (storageAvailable() && jwtToken) {
      const redirectUrl = localStorage.getItem('yp_redirect');
      localStorage.removeItem('yp_redirect');
      localStorage.removeItem('yp_provider');
      // After redirecting from Social Login, log to Segment
      const {
        sub: id,
        email,
        given_name: givenName,
        family_name: familyName,
      } = state.user?.attributes || {};
      segmentIdentifyAndTrack({
        event: 'Login',
        id,
        payload: {
          email,
          firstname: givenName,
          lastname: familyName,
        },
      });
      if (redirectUrl && redirectUrl !== 'undefined') {
        navigate(redirectUrl);
      }
    }
  }, [jwtToken, state]);

  // set listener for post social auth
  React.useEffect(() => {
    if (
      !Auth
    ) { return () => {}; }

    let isPending = false;
    const hubListener = async (data) => {
      const { payload: { event } } = data;

      if (event === 'signIn' && !isPending) {
        isPending = true;

        if (!state.user) {
          const currentUser = await Auth.currentAuthenticatedUser();
          dispatch({ type: 'AUTHENTICATED', user: currentUser });
        }
        isPending = false;
      }
    };

    Hub.listen('auth', hubListener);

    return () => {
      Hub.remove('auth', hubListener);
    };
  }, [Auth, state.user]);

  // refreshes user session when it expires
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    const now = new Date();
    const expiryDate = new Date(expiryTime * 1000);
    const delay = expiryDate - now;

    const timeout = setTimeout(async () => {
      refreshUser();
    }, delay);

    return () => { clearTimeout(timeout); };
  }, [expiryTime, refreshUser]);

  // replaces jwtToken when a customerId is selected
  React.useEffect(() => {
    if (!Auth) {
      return;
    }

    const getUserToken = async () => {
      if (
        isStaff
        && selectedCustomer?.customerId
        && selectedCustomer?.customerId !== customerId
      ) {
        const { username } = state.user;

        const challengedUser = await Auth.signIn(username);

        const user = await Auth.sendCustomChallengeAnswer(
          challengedUser,
          jwtToken,
          {
            contactId: selectedCustomer.contactId,
            customerId: selectedCustomer.customerId,
          },
        );

        if (getCustomerId({ user }) === selectedCustomer.customerId) {
          dispatch({ type: 'AUTHENTICATED', user });
        } else {
          setSelectedCustomer({ customerId });
          NotificationManager.error(
            `Customer ID ${selectedCustomer.customerId} couldn't be found in HubSpot. `
            + 'Please make sure you\'ve got the right one.',
          );
        }
      }
    };

    getUserToken();
  }, [Auth, customerId, isStaff, jwtToken, selectedCustomer, setSelectedCustomer, state.user]);

  authState = {
    // Auth Attributes
    state,
    jwtToken,
    payload,
    customerId,
    isStaff,

    // AWS Cognito Auth
    getCurrentUser,
    refreshUser,
    register,
    confirmSignUp,
    resendCode,
    login,
    logout,
    forgotPassword,
    forgotPasswordSubmit,
    changePassword,
    loginGoogle,
    loginFacebook,
    loginYellow,

    // Modal
    showHideModal,
    displayLoginModal,

    // Staff Tools
    getCustomerListByEmail,
    selectCustomer,
    selectedCustomer,
  };

  return (
    <AuthContext.Provider value={authState}>
      {children}
    </AuthContext.Provider>
  );
};

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
