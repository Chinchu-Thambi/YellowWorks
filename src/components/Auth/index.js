export { default, AuthProvider } from './AuthContext';
export { default as getCustomerId } from './services/getCustomerId';
