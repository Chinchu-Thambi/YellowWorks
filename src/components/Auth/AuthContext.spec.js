/* globals describe, it, jest, expect, Storage */

import React from 'react';
import { render, act } from '@testing-library/react/pure';
import Auth from '@aws-amplify/auth';

import AuthContext, { AuthProvider } from './AuthContext';

Auth.configure = jest.fn();

describe('AuthContext', () => {
  let authState;
  const jwtToken = 'jwtToken';
  const user = {
    attributes: {
      sub: 'sub',
      email: 'email@example.com',
      given_name: 'given_name',
      family_name: 'family_name',
    },
    signInUserSession: {
      idToken: {
        jwtToken,
      },
    },
  };
  const staffUser = {
    attributes: {
      sub: 'sub',
      email: 'email@yellow.co.nz',
      given_name: 'given_name',
      family_name: 'family_name',
    },
    signInUserSession: {
      idToken: {
        jwtToken,
        payload: {
          'cognito:groups': ['staff'],
        },
      },
    },
  };

  it('returns a null user for rejected auth', async () => {
    const rejectValue = 'user not logged in';

    await act(async () => {
      const currentAuthenticatedUser = jest.spyOn(Auth, 'currentAuthenticatedUser');
      currentAuthenticatedUser.mockRejectedValueOnce(rejectValue);

      render(
        <AuthProvider>
          <AuthContext.Consumer>
            {(latestState) => {
              authState = latestState;
            }}
          </AuthContext.Consumer>
        </AuthProvider>,
      );
    });

    expect(authState.state.status).toBe('UNAUTHENTICATED');
    expect(authState.state.user).toBeNull();
    expect(authState.state.error).toBe(rejectValue);
  });

  it('provides a user for successful auth', async () => {
    await act(async () => {
      const currentAuthenticatedUser = jest.spyOn(Auth, 'currentAuthenticatedUser');
      currentAuthenticatedUser.mockResolvedValueOnce(user);

      render(
        <AuthProvider>
          <AuthContext.Consumer>
            {(latestState) => {
              authState = latestState;
            }}
          </AuthContext.Consumer>
        </AuthProvider>,
      );
    });

    expect(authState.state.status).toBe('AUTHENTICATED');
    expect(authState.state.user).toBe(user);
  });

  it('displays a login modal', async () => {
    const target = '/target/url';
    const dismissTarget = '/dismissTarget/url';

    // Display Modal Action
    await act(async () => {
      authState.displayLoginModal({
        target, dismissTarget,
      });
    });

    expect(authState.state.modal.target).toBe(target);
    expect(authState.state.modal.dismissTarget).toBe(dismissTarget);
  });

  it('sets federated logins and redirects', async () => {
    const setItem = jest.spyOn(Storage.prototype, 'setItem');
    const federatedSignIn = jest.spyOn(Auth, 'federatedSignIn');

    setItem.mockClear();

    federatedSignIn.mockImplementation(() => {});

    await act(async () => {
      authState.loginGoogle();
      authState.loginFacebook();
      authState.loginYellow();
    });

    expect(setItem.mock.calls).toMatchSnapshot();
    expect(federatedSignIn.mock.calls).toMatchSnapshot();
  });

  it('provides login functionality', async () => {
    const username = 'username';
    const password = 'password';
    const signIn = jest.spyOn(Auth, 'signIn');
    signIn.mockResolvedValueOnce(user);
    let result;

    await act(async () => {
      result = await authState.login(username, password);
    });

    expect(signIn.mock.calls[0]).toEqual([username, password]);
    expect(result.user).toBe(user);
  });

  it('provides register functionality', async () => {
    const signUp = jest.spyOn(Auth, 'signUp');
    signUp.mockImplementation(() => {});

    const username = 'username';
    const password = 'password';
    const firstName = 'firstName';
    const lastName = 'lastName';

    await act(async () => {
      await authState.register(username, password, firstName, lastName);
    });

    expect(signUp.mock.calls[0][0]).toEqual({
      username,
      password,
      attributes: {
        given_name: firstName,
        family_name: lastName,
      },
    });
  });

  it('provides signup confirmation functionality', async () => {
    const resolvedValue = {};
    const confirmSignUp = jest.spyOn(Auth, 'confirmSignUp');
    confirmSignUp.mockResolvedValueOnce(resolvedValue);

    const username = 'username';
    const code = 'code';

    let result;
    await act(async () => {
      result = await authState.confirmSignUp(username, code);
    });

    expect(confirmSignUp.mock.calls[0]).toEqual([username, code]);
    expect(result).toEqual({
      success: true,
      result: resolvedValue,
    });
  });

  it('provides resend code functionality', async () => {
    const resendSignUp = jest.spyOn(Auth, 'resendSignUp');
    resendSignUp.mockResolvedValueOnce();

    const username = 'username';
    await act(async () => {
      await authState.resendCode(username);
    });

    expect(resendSignUp.mock.calls[0]).toEqual([username]);
  });

  it('provides password manipulation functionality', async () => {
    // forgot password
    const resolvedValue = {};
    const forgotPassword = jest.spyOn(Auth, 'forgotPassword');
    forgotPassword.mockResolvedValueOnce(resolvedValue);

    const username = 'username';
    const code = 'code';
    const newPassword = 'newPassword';

    let result;
    await act(async () => {
      result = await authState.forgotPassword(username);
    });

    expect(forgotPassword.mock.calls[0][0]).toBe(username);
    expect(result).toBe(resolvedValue);

    // reset forgotten password
    const forgotPasswordSubmit = jest.spyOn(Auth, 'forgotPasswordSubmit');
    forgotPasswordSubmit.mockResolvedValueOnce();

    await act(async () => {
      result = await authState.forgotPasswordSubmit(username, code, newPassword);
    });

    expect(forgotPasswordSubmit.mock.calls[0]).toEqual([
      username, code, newPassword,
    ]);

    // change password
    const changePassword = jest.spyOn(Auth, 'changePassword');
    changePassword.mockResolvedValueOnce(resolvedValue);
    const previousPassword = 'previousPassword';
    const proposedPassword = 'proposedPassword';

    await act(async () => {
      result = await authState.changePassword({ previousPassword, proposedPassword });
    });

    expect(result).toBe(resolvedValue);
    expect(changePassword.mock.calls[0]).toEqual([
      user, previousPassword, proposedPassword,
    ]);
  });

  it('provides logout functionality', async () => {
    const username = 'username';
    const password = 'password';
    const signOut = jest.spyOn(Auth, 'signOut');
    signOut.mockImplementation(() => {});

    await act(async () => {
      await authState.logout(username, password);
    });

    expect(signOut.mock.calls).toHaveLength(1);
    expect(authState.state.user).toBe(null);
  });

  it('uses current staff login to select a customerAccount', async () => {
    const customerObject = {
      customerId: 'customerId',
    };

    const challengedUser = {};

    Auth.signIn = jest.fn();
    Auth.sendCustomChallengeAnswer = jest.fn();

    await act(async () => {
      Auth.signIn.mockResolvedValueOnce(staffUser);
      await authState.login();

      Auth.signIn.mockResolvedValueOnce(challengedUser);
      Auth.sendCustomChallengeAnswer.mockResolvedValueOnce(user);
      await authState.selectCustomer(customerObject);
    });

    expect(Auth.sendCustomChallengeAnswer.mock.calls[0][0]).toEqual(
      challengedUser,
      jwtToken,
      { customerId: customerObject.customerId },
    );

    Auth.signIn.mockRestore();
    Auth.sendCustomChallengeAnswer.mockRestore();
  });
});
