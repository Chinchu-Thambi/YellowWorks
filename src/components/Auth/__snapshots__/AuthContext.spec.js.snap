// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthContext sets federated logins and redirects 1`] = `
Array [
  Array [
    "__storage_test__",
    "__storage_test__",
  ],
  Array [
    "yp_redirect",
    "/target/url",
  ],
  Array [
    "yp_provider",
    "Google",
  ],
  Array [
    "__storage_test__",
    "__storage_test__",
  ],
  Array [
    "yp_redirect",
    "/target/url",
  ],
  Array [
    "yp_provider",
    "Facebook",
  ],
  Array [
    "__storage_test__",
    "__storage_test__",
  ],
  Array [
    "yp_redirect",
    "/target/url",
  ],
  Array [
    "yp_provider",
    "Yellow",
  ],
]
`;

exports[`AuthContext sets federated logins and redirects 2`] = `
Array [
  Array [
    Object {
      "provider": "Google",
    },
  ],
  Array [
    Object {
      "provider": "Facebook",
    },
  ],
  Array [
    Object {
      "provider": "Yellow",
    },
  ],
]
`;
